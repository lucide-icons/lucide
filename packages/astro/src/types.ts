import type { HTMLAttributes } from 'astro/types';

// Type that the Astro language server needs to infer component props in Astro files
export type AstroComponent = (_props: IconProps) => any;

export interface IconProps extends SVGAttributes {
  color?: string;
  size?: number | string;
  'stroke-width'?: number | string;
  absoluteStrokeWidth?: boolean;
  class?: string;
  iconNode?: IconNode;
}

export type SVGAttributes = HTMLAttributes<'svg'>;

export type IconNode = IconNodeChild[];

type IconNodeChild = [elementName: SVGElements, attrs: SVGAttributes];

// All possible svg elements according to the Astro definitions
type SVGElements =
  | 'svg'
  | 'animate'
  | 'circle'
  | 'clipPath'
  | 'defs'
  | 'desc'
  | 'ellipse'
  | 'feBlend'
  | 'feColorMatrix'
  | 'feComponentTransfer'
  | 'feComposite'
  | 'feConvolveMatrix'
  | 'feDiffuseLighting'
  | 'feDisplacementMap'
  | 'feDistantLight'
  | 'feFlood'
  | 'feFuncA'
  | 'feFuncB'
  | 'feFuncG'
  | 'feFuncR'
  | 'feGaussianBlur'
  | 'feImage'
  | 'feMerge'
  | 'feMergeNode'
  | 'feMorphology'
  | 'feOffset'
  | 'fePointLight'
  | 'feSpecularLighting'
  | 'feSpotLight'
  | 'feTile'
  | 'feTurbulence'
  | 'filter'
  | 'foreignObject'
  | 'g'
  | 'image'
  | 'line'
  | 'linearGradient'
  | 'marker'
  | 'mask'
  | 'metadata'
  | 'path'
  | 'pattern'
  | 'polygon'
  | 'polyline'
  | 'radialGradient'
  | 'rect'
  | 'stop'
  | 'switch'
  | 'symbol'
  | 'text'
  | 'textPath'
  | 'tspan'
  | 'use'
  | 'view';
