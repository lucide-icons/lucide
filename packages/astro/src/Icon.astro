---
import defaultAttributes from './defaultAttributes';
import type { IconProps as Props } from './types';

const {
  color = 'currentColor',
  size = 24,
  'stroke-width': strokeWidth = 2,
  absoluteStrokeWidth = false,
  iconNode = [],
  class: className,
  ...rest
} = Astro.props;
---

<svg
  {...{
    ...defaultAttributes,
    width: size,
    height: size,
    stroke: color,
    'stroke-width': absoluteStrokeWidth ? (Number(strokeWidth) * 24) / Number(size) : strokeWidth,
    ...rest,
  }}
  class:list={['lucide', className]}
>
  {iconNode.map(([Tag, attrs]) => <Tag {...attrs} />)}
  <slot />
</svg>
