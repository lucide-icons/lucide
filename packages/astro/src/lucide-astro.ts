export * from './icons/index';
export * as icons from './icons/index';
export * from './aliases';
export * from './types';

export { default as defaultAttributes } from './defaultAttributes';
export { default as createLucideIcon } from './createLucideIcon';
export { default as Icon } from './Icon.astro';
