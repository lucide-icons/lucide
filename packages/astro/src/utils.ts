/**
 * Converts string to kebab case
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string): string =>
  string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

/**
 * Merges classes into a single string
 *
 * @param {array} classes
 * @returns {string} A string of classes
 */
export const mergeClasses = <ClassType = string | undefined | null>(
  ...classes: ClassType[]
): string =>
  classes
    .filter((className, index, array) => {
      return (
        Boolean(className) &&
        (className as string).trim() !== '' &&
        array.indexOf(className) === index
      );
    })
    .join(' ')
    .trim();
