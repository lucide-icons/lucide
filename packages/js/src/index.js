import icons from './icons';
import toSvg from './to-svg';
import replace from './replace';

module.exports = { icons, toSvg, replace };
