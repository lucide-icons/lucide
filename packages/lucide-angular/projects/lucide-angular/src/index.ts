
export * from './lib/lucide-angular.component'
export * from './lib/lucide-angular.module';
export * from './helpers/create-element'

export * from './icons'
export * as icons from './icons';
