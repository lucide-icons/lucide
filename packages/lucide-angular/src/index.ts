import * as icons from './icons';

export * from './lib/lucide-angular.component';
export * from './lib/lucide-angular.module';
export * from './helpers/create-element';
export * from './icons';
export { icons };
