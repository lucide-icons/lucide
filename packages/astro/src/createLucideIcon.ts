import { mergeClasses, toKebabCase } from './utils';
import type { AstroComponentFactory } from 'astro/runtime/server/render/astro/factory.js';
import type { IconNode } from './types';
import { render, renderSlot, createComponent, renderComponent } from 'astro/compiler-runtime';
import Icon from './Icon.astro';

export default (iconName: string, iconNode: IconNode): AstroComponentFactory => {
  const Component = createComponent(
    ($$result, $$props: Record<string, any>, $$slots) => {
      const { class: className, ...restProps } = $$props;
      return render`${renderComponent(
        $$result,
        'Icon',
        Icon,
        {
          class: mergeClasses(
            Boolean(iconName) && `lucide-${toKebabCase(iconName)}`,
            Boolean(className) && className,
          ),
          iconNode,
          ...restProps,
        },
        { default: () => render`${renderSlot($$result, $$slots['default'])}` },
      )}`;
    },
    undefined,
    'none',
  );
  return Component;
};
