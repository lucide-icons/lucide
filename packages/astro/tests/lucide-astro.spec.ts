import { describe, it, expect } from 'vitest';
import { Pen, Edit2, Grid, Droplet, Smile } from '../src/lucide-astro';
import defaultAttributes from '../src/defaultAttributes';
import { createAstroHTMLString, render } from './utils';

describe('Using lucide icon components', () => {
  it('should render a component', async () => {
    const { container } = await render(Grid);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should render the icon with default attributes', async () => {
    const { container } = await render(Grid);
    const SVGElement = container.firstElementChild;

    expect(SVGElement).toHaveAttribute('xmlns', defaultAttributes.xmlns);
    expect(SVGElement).toHaveAttribute('width', String(defaultAttributes.width));
    expect(SVGElement).toHaveAttribute('height', String(defaultAttributes.height));
    expect(SVGElement).toHaveAttribute('viewBox', defaultAttributes.viewBox);
    expect(SVGElement).toHaveAttribute('fill', defaultAttributes.fill);
    expect(SVGElement).toHaveAttribute('stroke', defaultAttributes.stroke);
    expect(SVGElement).toHaveAttribute('stroke-width', String(defaultAttributes['stroke-width']));
    expect(SVGElement).toHaveAttribute('stroke-linecap', defaultAttributes['stroke-linecap']);
    expect(SVGElement).toHaveAttribute('stroke-linejoin', defaultAttributes['stroke-linejoin']);

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should adjust the size, stroke color and stroke width', async () => {
    const { container } = await render(Grid, {
      props: {
        size: 48,
        stroke: 'red',
        'stroke-width': 4,
      },
    });

    const SVGElement = container.firstElementChild;

    expect(SVGElement).toHaveAttribute('stroke', 'red');
    expect(SVGElement).toHaveAttribute('width', '48');
    expect(SVGElement).toHaveAttribute('stroke-width', '4');

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should render the alias icon', async () => {
    const { container: PenIconContainer } = await render(Pen, {
      props: {
        size: 48,
        stroke: 'red',
        'stroke-width': 4,
      },
    });

    const { container: Edit2Container } = await render(Edit2, {
      props: {
        size: 48,
        stroke: 'red',
        'stroke-width': 4,
      },
    });

    expect(PenIconContainer.innerHTML).toBe(Edit2Container.innerHTML);
  });

  it('should not scale the strokeWidth when absoluteStrokeWidth is set', async () => {
    const { container } = await render(Grid, {
      props: {
        size: 48,
        stroke: 'red',
        absoluteStrokeWidth: true,
      },
    });

    const SVGElement = container.firstElementChild;

    expect(SVGElement).toHaveAttribute('stroke', 'red');
    expect(SVGElement).toHaveAttribute('width', '48');
    expect(SVGElement).toHaveAttribute('height', '48');
    expect(SVGElement).toHaveAttribute('stroke-width', '1');

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should add a non-default attribute to the element', async () => {
    const { container } = await render(Smile, {
      props: {
        style: 'position: absolute',
      },
    });

    const SVGElement = container.firstElementChild;

    expect(SVGElement).toHaveAttribute('style', 'position: absolute');

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should pass children to the icon slot', async () => {
    const { getByText, container } = await render(Smile, {
      slots: {
        default: createAstroHTMLString('<p>Hello World</p>'),
      },
    });

    const textElement = getByText('Hello World');

    expect(textElement).toBeInTheDocument();

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should apply all classes to the element', async () => {
    const { container } = await render(Droplet, {
      props: {
        class: 'my-icon',
      },
    });
    const SVGElement = container.firstElementChild;

    expect(SVGElement).toHaveClass('my-icon');
    expect(SVGElement).toHaveClass('lucide');
    expect(SVGElement).toHaveClass('lucide-droplet');

    expect(container.innerHTML).toMatchSnapshot();
  });
});
