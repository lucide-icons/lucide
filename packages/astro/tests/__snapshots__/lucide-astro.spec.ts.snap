// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using lucide icon components > should add a non-default attribute to the element 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     style="position: absolute"
     class="lucide lucide-smile"
>
  <circle cx="12"
          cy="12"
          r="10"
  >
  </circle>
  <path d="M8 14s1.5 2 4 2 4-2 4-2">
  </path>
  <line x1="9"
        x2="9.01"
        y1="9"
        y2="9"
  >
  </line>
  <line x1="15"
        x2="15.01"
        y1="9"
        y2="9"
  >
  </line>
</svg>
`;

exports[`Using lucide icon components > should adjust the size, stroke color and stroke width 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="48"
     height="48"
     viewbox="0 0 24 24"
     fill="none"
     stroke="red"
     stroke-width="4"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-grid-3x3"
>
  <rect width="18"
        height="18"
        x="3"
        y="3"
        rx="2"
  >
  </rect>
  <path d="M3 9h18">
  </path>
  <path d="M3 15h18">
  </path>
  <path d="M9 3v18">
  </path>
  <path d="M15 3v18">
  </path>
</svg>
`;

exports[`Using lucide icon components > should apply all classes to the element 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-droplet my-icon"
>
  <path d="M12 22a7 7 0 0 0 7-7c0-2-1-3.9-3-5.5s-3.5-4-4-6.5c-.5 2.5-2 4.9-4 6.5C6 11.1 5 13 5 15a7 7 0 0 0 7 7z">
  </path>
</svg>
`;

exports[`Using lucide icon components > should not scale the strokeWidth when absoluteStrokeWidth is set 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="48"
     height="48"
     viewbox="0 0 24 24"
     fill="none"
     stroke="red"
     stroke-width="1"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-grid-3x3"
>
  <rect width="18"
        height="18"
        x="3"
        y="3"
        rx="2"
  >
  </rect>
  <path d="M3 9h18">
  </path>
  <path d="M3 15h18">
  </path>
  <path d="M9 3v18">
  </path>
  <path d="M15 3v18">
  </path>
</svg>
`;

exports[`Using lucide icon components > should pass children to the icon slot 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-smile"
>
  <circle cx="12"
          cy="12"
          r="10"
  >
  </circle>
  <path d="M8 14s1.5 2 4 2 4-2 4-2">
  </path>
  <line x1="9"
        x2="9.01"
        y1="9"
        y2="9"
  >
  </line>
  <line x1="15"
        x2="15.01"
        y1="9"
        y2="9"
  >
  </line>
  <p>
    Hello World
  </p>
</svg>
`;

exports[`Using lucide icon components > should render a component 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-grid-3x3"
>
  <rect width="18"
        height="18"
        x="3"
        y="3"
        rx="2"
  >
  </rect>
  <path d="M3 9h18">
  </path>
  <path d="M3 15h18">
  </path>
  <path d="M9 3v18">
  </path>
  <path d="M15 3v18">
  </path>
</svg>
`;

exports[`Using lucide icon components > should render the icon with default attributes 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-grid-3x3"
>
  <rect width="18"
        height="18"
        x="3"
        y="3"
        rx="2"
  >
  </rect>
  <path d="M3 9h18">
  </path>
  <path d="M3 15h18">
  </path>
  <path d="M9 3v18">
  </path>
  <path d="M15 3v18">
  </path>
</svg>
`;
